test/integration/sessions/sessions.test.ts

```diff
@@ -72,93 +73,108 @@ describe('Sessions Spec', function () {
-          {
-            description: 'should support operations that throw exceptions',
-            operation: (/* client */) => (/* session */) => {
-              throw new Error('something went wrong!');
-            }
+      const tests = [
+        {
+          description: 'should resolve non-async callbacks that return promises',
+          operation: client => session => {
+            return client.db('test').collection('foo').find({}, { session }).toArray();
           }
-        ].forEach(testCase => {
-          it(testCase.description, function () {
-            const client = test.client;
-
-            return client
-              .withSession(testCase.operation(client))
-              .then(
-                () => expect(client.topology.s.sessionPool.sessions).to.have.length(1),
-                () => expect(client.topology.s.sessionPool.sessions).to.have.length(1)
-              )
-              .then(() => client.close())
-              .then(() => {
-                // verify that the `endSessions` command was sent
-                const lastCommand = test.commands.started[test.commands.started.length - 1];
-                expect(lastCommand.commandName).to.equal('endSessions');
-                expect(client.topology).to.not.exist;
-              });
-          });
-        });
+        },
+        {
+          description: 'should resolve async callbacks',
+          operation: client => async session =>
+            await client.db('test').collection('foo').find({}, { session }).toArray()
+        },
+        {
+          description: 'should reject with error thrown from async callback',
+          operation: (/* client */) => async (/* session */) => {
+            throw new Error('thrown from async function');
+          }
+        },
+        {
+          description: 'should reject callbacks that return a rejected promise',
+          operation: (/* client */) => (/* session */) => {
+            return Promise.reject(new Error('something awful'));
+          }
+        },
+        {
+          description: 'should resolve callbacks that do not return a promise',
+          operation: (/* client */) => (/* session */) => {
+            // This is incorrect usage of the API, but we're making sure that we don't use
+            // .then on the result of the callback, we should always start with a Promise.resolve()
+            //
+            // void return;
+          }
+        },
+        {
+          description: 'should reject callbacks that throw synchronous exceptions',
+          operation: (/* client */) => (/* session */) => {
+            throw new Error('something went wrong!');
+          }
+        }
+      ];
 
-        it('supports passing options to ClientSession', function () {
-          const client = test.client;
+      for (const testCase of tests) {
+        it(testCase.description, async function () {
+          const shouldResolve = testCase.description.startsWith('should resolve');
```

Comment:
This isn't blocking - but could we just add a property to the tests instead to indicate whether or not the test should resolve or reject?

```typescript
{
      description: 'should reject callbacks that throw synchronous exceptions',
      operation: (/* client */) => (/* session */) => {
        throw new Error('something went wrong!');
      },
      result: 'resolve' // or something like this
}

...

const shoudResolve = result === 'resolve';
```